### PR TITLE
Hdf5

### PR DESCRIPTION
**Pull request by justusschock**
## Short Description
This should add a simple way to convert a dataset to hdf5 for faster loading.
It includes a dependency on h5py, which is only called, when the actual functions are called, so I probably would only add this to the test requirements.

Regarding the tests: So far there are none. This is because I am not sure on how to test this. I could convert a rather small dataset (like mnist) to hdf5 and then compare the results, but this would probably slow down our tests substantially.
Any ideas?

### PR Implementer
This is a small checklist for the implementation details of this PR.
If you submit a PR, please look at these points (don't worry about the `RisingTeam`
and `Reviewer` workflows, the only purpose of those is to have a compact view of
the steps)

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/PhoenixDL/rising/blob/master/CONTRIBUTING.md).

- [x] Implementation
- [x] Docstrings & Typing
- [ ] Check `__all__` sections and `__init__`
- [ ] Unittests (look at the line coverage of your tests, the goal is 100%!)
- [ ] Update notebooks & documentation if necessary
- [ ] Pass all tests
- [ ] Add the checksum of the last implementation commit to the Changelog

### RisingTeam
<details>
  <summary>RisingTeam workflow</summary>

  - [ ] Add pull request to project (optionally delete corresponding project note)
  - [ ] Assign correct label (if you don't have permission to do this, someone will do it for you. 
      Please make sure to communicate the current status of the pr.)
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)
</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>
  
  - [ ] Do all tests pass? (Unittests, NotebookTests, Documentation)
  - [ ] Does the implementation follow `rising` design conventions?
  - [ ] Are the tests useful? (!!!) Are additional tests needed? 
        Can you think of critical points which should be covered in an additional test?
  - [ ] Optional: Check coverage locally / Check tests locally if GPU is necessary to execute
</details>

### CODE REVIEW

File: rising/loading/hdf5.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        -------
+        dict
+            a dictionary containing one sample per key in the dataset
+        """
+
+        import h5py
+        if isinstance(self.hdf5_path, str):
+            file = h5py.File(self.hdf5_path, 'r')
+        else:
+            file = self.hdf5_path
```
Comment (mibaumgartner) [inline rising/loading/hdf5.py]: Can we add support for `pathlib.Path` here?

Comment (mibaumgartner) [inline rising/loading/hdf5.py]: Can we add support for `pathlib.Path` here?

File: rising/loading/hdf5.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        data = file.create_group(dset_name)
+
+        keys_to_skip = []
+        dsets = {}
+
+        for idx in tqdm(range(0, len(dset))):
+            sample = numpy_collate(dset[idx])
+
+            if isinstance(sample, Sequence):
+                sample = {idx, v for idx, v in enumerate(sample)}
```
Comment (mibaumgartner) [inline rising/loading/hdf5.py]: Is this really intended or should this be `{idx: v ....}`

File: rising/loading/hdf5.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
+from rising.loading.dataset import Dataset
+import os
+import numpy as np
+from tqdm import tqdm
+
+from rising.loading.collate import numpy_collate
+
+
+def h5ify(dset: Dataset, hdf5_path: str, dset_name: str, overwrite: bool = False,
```
Comment (mibaumgartner) [inline rising/loading/hdf5.py]: Nit: Can we make this function shorter or extract some general part into a second function? It is a bit long for my liking ^^

Comment (justusschock) [reply rising/loading/hdf5.py]: I wouldn't split this, because there is no need to. We most likely don't need this code anywhere else...

**Comment by mibaumgartner** [comment]
Datasets will be moved to a different repo.

**Pull request closed without merging**